include/dali/core/exec/tasking/sync.h

```diff
@@ -0,0 +1,180 @@
+
+  /** Checks whether the Waitable is ready to be acquired */
+  virtual bool CheckComplete() const = 0;
+
+  /** Changes the state of the object to acquired (if necessary) and returns whether the operation
+   *  was successful.
+   */
+  virtual bool AcquireImpl() = 0;
+
+  void Notify(Scheduler &sched);
+
+  static bool task_owner_equal(const WeakTask &w, const SharedTask &t) {
+    return !w.owner_before(t) && !t.owner_before(w);
+  }
+
+  bool IsWaitedForBy(const SharedTask &task) const {
+    auto it = std::find_if(waiting_.begin(), waiting_.end(),
+                           [&](auto &t) { return task_owner_equal(t, task); });
+    return it != waiting_.end();
+  }
+
+  /** Tries to acquire the waitable object on behalf of a task
+   *
+   * This function can fail for two reasons:
+   * - the task is not waiting for this waitable
+   * - AcquireImpl fails
+   */
+  bool TryAcquire(const SharedTask &task) {
+    auto it = std::find_if(waiting_.begin(), waiting_.end(),
+                           [&](auto &t) { return task_owner_equal(t, task); });
+    if (it == waiting_.end())
+      return false;
+    if (AcquireImpl()) {
+      waiting_.erase(it);
+      return true;
+    }
+    return false;
+  }
+
+  bool AddToWaiting(const SharedTask &task) {
+    if (IsWaitedForBy(task))
+      return false;
+    waiting_.push_back(task);
+    return true;
+  }
+
+ public:
+  virtual ~Waitable() = default;
+};
+
+/** A waitable object that changes the state at most once, from incomplete to complete.
+ *
+ * A completion event is any Waitable which cannot become "unsignalled" - for example, once
+ * a task is complete it will ever remain so, it cannot go back to the incomplete state.
+ */
+class CompletionEvent : public Waitable {
+ protected:
+  bool AcquireImpl() override {
+    // Nothing to acquire - just return true if the event is completed.
+    return CheckComplete();
+  }
+
+  void MarkAsComplete() {
+    completed_ = true;
+  }
+
+  bool CheckComplete() const override {
+    return completed_;
+  }
+
+ private:
+  bool completed_ = false;
+};
+
+/** A waitable object which has a Release method, which can be called from outside any task.
+ */
+class Releasable : public Waitable {
+ public:
+  bool Release(Scheduler &sched) {
+    if (!ReleaseImpl())
+      return false;
+    Notify(sched);
+    return true;
+  }
+
+ protected:
+  /** Changes the internal state of the object.
+   *
+   * If CheckComplete is called atomically after ReleaseImpl, it must return true.
+   */
+  virtual bool ReleaseImpl() = 0;
+};
+
+/** A releasable object which counts how many times it can be acquired.
+ */
+class Semaphore : public Releasable {
+ public:
+  explicit Semaphore(int max_count) : Semaphore(max_count, max_count) {}
+  Semaphore(int max_count, int initial_count) : count(initial_count), max_count(max_count) {}
+
+ protected:
+  mutable spinlock lock_;
+
+  bool CheckComplete() const override {
+    std::lock_guard g(lock_);
+    return count > 0;
```

Comment:
I'm not sure if I understand the semantics here. Should it compare with 0 or `max_count`?

Reply:
Complete means that it can admit a task. Arguably that's not the best name for waitables other than CompletionEvent. If you can think of a better name for the condition when a Waitable can be acquired, please suggest. I didn't want to use Ready, because it's already quite confusing with tasks (a task is eeady when it can be executed, not when it's done).

Reply:
Maybe `IsAcquirable`?

Reply:
Yeah, probably better.

Reply:
Changed.